### DISCUSSION TITLE
增加挂机完成通知

### DISCUSSION THREAD

**Discussion by Higher-Stark** [MaaAssistantArknights/MaaAssistantArknights#4322] [Ideas]
目前使用MAA有遇到过这样的情况：
1. 模拟器运行一段时间后不稳定，可能会挂掉，蓝叠模拟器有时奔溃后可能有残留进程导致无法重启模拟器
2. 上午MAA公招出了高资，当天10点大版本更新+刷新公招tag

当人不在电脑旁边，就看不到这些情况，所以最好给MAA加一个通知机制。

最近的版本MAA加入了运行前脚本和运行后脚本，利用这个自己写了一点batch和python脚本，实现运行完成后把新增的日志（gui.log）以邮件的方式发送出去，手机上查看邮件就能看到详细的运行过程。
不过今天更新了v4.14.1发现MAA在关闭之后才会把日志写进gui.log，自己的脚本没办法在MAA运行的时候从gui.log中得到新的日志。

所以希望开发者能给MAA增加发送通知的功能。
像近期刷1-7，MAA的日志会很长，所以个人建议优先实现邮件的通知方式，后续增加公众号推送、钉钉机器人、telegram bot等通知方式。

**Comment by Higher-Stark** [comment]
分享一下目前自己的邮件通知脚本。
设备环境: Windows 10, Python 3.10.5
首先说一下目录，我是在MAA的顶层目录下建了一个notification文件夹，把后面的脚本都放在了notification这个文件夹里了。
首先是开始前脚本`preTask.bat`，就是记录一下当前`gui.log`的日志长度。
```batch
@echo off

pushd <path to MAA top directory> 
find /c /v "" debug\gui.log > debug\lines_1.tmp

popd
```
然后是结束后脚本`postTask.bat`，记录一下日志更新后的长度，然后把前后两次的日志长度告诉`mail.py`，让`mail.py`把新的日志通过邮件发送出去。
```batch
@echo off
setlocal enabledelayedexpansion
pushd <path to MAA top directory> 

timeout /t 3 /nobreak

find /c /v "" debug\gui.log > debug\lines_2.tmp

SET BIN=<path of python.exe>

for /f "tokens=3" %%a in (debug\lines_1.tmp) do (
    set line1=%%a & goto :continue1
)
:continue1

for /f "tokens=3" %%a in (debug\lines_2.tmp) do (
    set line2=%%a & goto :continue2
)
:continue2

@REM Enter notification directory
cd notification

echo "%BIN% mail.py %line1% %line2%" >> notification.log
%BIN% mail.py %line1% %line2%

cd ..

del debug\lines_1.tmp
del debug\lines_2.tmp

popd
```
接下来是`mail.py`，这里有几个依赖：jinja2, dynaconf。dynaconf用来生成配置文件；jinja2是根据日志和模板文件来生成html页面的，不是必需的，可以改脚本删掉生成html的部分，直接发纯文本。
```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import sys
import linecache
import logging as log
from config import settings

def render_template(template, **kwargs):
    ''' renders a Jinja template into HTML '''
    # check if template exists
    if not os.path.exists(template):
        print('No template file present: %s' % template)
        sys.exit()

    import jinja2
    templateLoader = jinja2.FileSystemLoader(searchpath=".")
    templateEnv = jinja2.Environment(loader=templateLoader, trim_blocks=True, lstrip_blocks=True)
    templ = templateEnv.get_template(template)
    return templ.render(**kwargs)
#------------------------------------------------------------------------------------------------
def send_email(to, sender='MyCompanyName<noreply@mycompany.com>', cc=None, bcc=None, subject=None, body=None):
    ''' sends email using a Jinja HTML template '''
    import smtplib
    # Import the email modules
    from email.mime.multipart import MIMEMultipart
    from email.mime.text import MIMEText
    from email.header import Header
    from email.utils import formataddr

    user = settings.user
    password = settings.password

    # convert TO into list if string
    if type(to) is not list:
        to = to.split()

    to_list = to + [cc] + [bcc]
    to_list = list(filter(lambda x: x is not None, to_list)) # remove null emails

    msg = MIMEMultipart('alternative')
    msg['From']    = sender
    msg['Subject'] = subject
    msg['To']      = ','.join(to_list)
    msg['Cc']      = cc
    msg['Bcc']     = bcc
    msg.attach(MIMEText(body, 'html', 'utf8'))
    server = smtplib.SMTP(settings.smtp_server) # or your smtp server
    try:
        log.info('sending email {}'.format(subject))
        server.login (user=user, password=password)
        server.sendmail(sender, to_list, msg.as_string())
    except Exception as e:
        log.error('Error sending email')
        log.exception(str(e))
    finally:
        server.quit()
#------------------------------------------------------------------------------------------------

# MAIN

start = int(sys.argv[1])
end = int(sys.argv[2])
content = linecache.getlines(settings.log_file)[start: end]

# generate HTML from template
html = render_template('template.j2', **locals())

to_list = list(settings.recipients)
sender = settings.sender
cc = settings.cc
bcc = settings.bcc
subject = settings.subject

# send email to a list of email addresses
send_email(to_list, sender, cc, bcc, subject, html)
```

`mail.py`的配置文件`settings.toml`示例：
```toml
recipients = ['YOUR@MAIL']
sender = 'MAA Bot <YOUR@MAIL>'
cc = 'YOUR@MAIL'
bcc = ''
subject = 'MAA任务完成通知'
log_file = '../debug/gui.log'
smtp_server = '<smtp server address>'
```
在.secret.toml里面写自己邮箱的账号和密码：
```
user = 'YOUR@MAIL'
password = '**************'
```

jinja2模板`template.j2`
```jinja
<style type="text/css">
* {
  margin: 0;
  padding: 0;
  /* Fix CSS. */
  box-sizing: border-box;
}

body {
  margin: 20px;
}

/* Set font size and color for code blocks */
code {
  font-size: 14px;
  color: #333;
  font-family: "Noto Sans CJK", "SimSun", "Microsoft JhengHei", sans-serif;
  display: block;
  white-space: pre-wrap;
}

/* Set background color and padding for pre element */
pre {
  background-color: #f5f5f5;
  padding: 10px;
  border: 1px solid #ccc;
  border-radius: 5px;
  overflow-x: auto; /* Add horizontal scrollbar if content is wider than the container */
}

img {
  max-width: 100%;
  height: auto;
}

/* Set font size and line height for pre element */
pre code {
  font-size: 14px;
  line-height: 1.5;
}

a {
  font-size: 12px;
  line-height: 1.0;
  color: #D3D3D3;
}

#maa {
  width: 80%;
  height: auto;
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.center {
  text-align: center;
}
</style>
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <!-- Link to Bootstrap CSS -->
  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css" integrity="sha384-HyQr4qM6g6tNmNJiRBmZfH3bx9n0f/CbbVcwS5xgPpV/WmG1aXfZtYyyJ0MImk3X" crossorigin="anonymous">
  <script src="https://code.jquery.com/jquery-3.5.1.slim.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/@popperjs/core@2.9.3/dist/umd/popper.min.js"></script>
  <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js"></script>
  <script>
      window.onload = function () {
        document.querySelector("#header_img").onclick = function () {
          document.querySelector("#bottom").scrollIntoView({ alignToTop: false, behavior: "smooth" });
        }
      }
  </script>
</head>
<body>
  <header>
    <img id="header_img" src="https://images4.alphacoders.com/103/1034459.jpg" alt="Header Image">
    <!--h1>Welcome to my website</h1-->
    <h2 class="center">MAA GUI日志</h2>
  </header>
  <!--div><a href="#bottom">跳转底部</a></div-->
  <pre><code>
{% for line in content %}
  {# line | replace("\r\n", "") #}
{{ line | replace("\n", "") }}
{% endfor %}
</code></pre>
<div id="bottom" class="footer">
  <img id="maa" src="https://cdn.jsdelivr.net/gh/MaaAssistantArknights/design/banner/maa-banner_original.jpg" alt="MAA Icon">
</div>
</body>
```

贴一个手机端通知邮件的效果
![MAA通知邮件示例](https://user-images.githubusercontent.com/25293291/232307553-e78fada4-6f30-4d09-b2e2-dd1aa5bed1bf.jpg)

**Comment by MistEO** [comment]
@LiamSho gui log 能手动 sync 吗？感觉可以在收到部分回调的时候 sync 一下

**Comment by AlisaAkiron** [comment]
要不干脆直接集成一下 ServerChan 好了（

**Reply by Yebolin** [reply]
为什么不是通用webhook。这样可以自由配置。

**Comment by sappho192** [comment]
Hi, I've been using MAA around 6 months and recently I want to add nearly same feature in this discussion: "Send an e-mail when all the task is finished".
@Higher-Stark do you have any progress about this feature?
Or it would be also nice if there're callback or function which returns current MAA's task status, so our own program will just periodically check MAA's status and send and e-mail if all the task is finished.

===========
翻译：DeepL：
你好，我使用 MAA 已经有 6 个月了，最近我想添加与本次讨论中几乎相同的功能： "任务完成后发送电子邮件"。
@Higher-Stark，您对这项功能有什么进展吗？
如果有回调或函数返回当前 MAA 的任务状态就更好了，这样我们自己的程序就可以定期检查 MAA 的状态，并在所有任务完成后发送电子邮件。

**Reply by sappho192** [reply]
I've made a simple Windows script which can send an e-mail after the task finished.

[maa-mailerscript.zip](https://github.com/MaaAssistantArknights/MaaAssistantArknights/files/12274109/maa-mailerscript.zip)
You just need two script files in above zip file:
## mailer.bat
```
@echo off
powershell -ExecutionPolicy Bypass -File ".\send_email.ps1"
```
## send_email.ps1
```
$From = "YOUR_ID@gmail.com"
$To = "YOUR_ID@gmail.com"
$Subject = "MAA notification"
$Body = "MAA task is finished!"

# The password is an app-specific password if you have 2-factor-auth enabled
$Password = "YOUR_PASSWORD" | ConvertTo-SecureString -AsPlainText -Force
$Credential = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $From, $Password
Send-MailMessage -From $From -To $To -Subject $Subject -Body $Body -SmtpServer "smtp.gmail.com" -port 587 -UseSsl -Credential $Credential
```
Since Windows PowerShell provides `Send-MailMessage` which can simply send an e-mail, you can send an e-mail by executing the script!  
Like following screenshot, you can just add the location of `mailer.bat` and it will work. (Of course you need to write in your information in `send_email.ps1`.

![04-2](https://github.com/MaaAssistantArknights/MaaAssistantArknights/assets/7788738/c51de961-24a1-409a-bcf1-87a5f25c10b9)
![06](https://github.com/MaaAssistantArknights/MaaAssistantArknights/assets/7788738/551e220d-dc79-498f-b0e2-5a445b2c7eb7)

I'm going to add a feature to do this inside MAA GUI app.

=====译文 DeepL
我制作了一个简单的 Windows 脚本，可以在任务完成后发送电子邮件。
[maa-mailerscript.zip](https://github.com/MaaAssistantArknights/MaaAssistantArknights/files/12274109/maa-mailerscript.zip)
您只需要上述压缩文件中的两个脚本文件：
## mailer.bat
```
@echo off
powershell -ExecutionPolicy Bypass -File ".\send_email.ps1"
```
## send_email.ps1
```
$From = "YOUR_ID@gmail.com"
$To = "YOUR_ID@gmail.com"
$Subject = "MAA notification"
$Body = "MAA task is finished!"

# The password is an app-specific password if you have 2-factor-auth enabled
$Password = "YOUR_PASSWORD" | ConvertTo-SecureString -AsPlainText -Force
$Credential = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $From, $Password
Send-MailMessage -From $From -To $To -Subject $Subject -Body $Body -SmtpServer "smtp.gmail.com" -port 587 -UseSsl -Credential $Credential
```

由于 Windows PowerShell 提供的 `Send-MailMessage` 可以简单地发送电子邮件，因此你可以通过执行脚本来发送电子邮件！ 
就像下面的截图一样，你只需添加`mailer.bat`的位置就可以了。(当然，你需要在 `send_email.ps1` 中写入你的信息。

![04-2](https://github.com/MaaAssistantArknights/MaaAssistantArknights/assets/7788738/c51de961-24a1-409a-bcf1-87a5f25c10b9)
![06](https://github.com/MaaAssistantArknights/MaaAssistantArknights/assets/7788738/551e220d-dc79-498f-b0e2-5a445b2c7eb7)

我将在 MAA GUI 应用程序中添加一项功能来实现这一点。

**Reply by MistEO** [reply]
You can read `debug/gui.json` with a filter to add more information to the email!

**Reply by sappho192** [reply]
> You can read `debug/gui.json` with a filter to add more information to the email!

@MistEO Thank you for your advice. :)
Actually, I've already looked into `debug/gui.log` but I think it's very hard work to filter out this log file to get dedicated information.
For example, you can see that log data is not structured and I need to consider multi-language stuff if I want to extract "Which operation is finished" or "How many RMA-70-12 are acquired".
![image](https://github.com/MaaAssistantArknights/MaaAssistantArknights/assets/7788738/f3b6ec6c-a3e4-4309-9634-646c66035136)
Also, if I run multiple MAA & app player, it will be challenging to distinguish each MAA instance.

But at least I can try inserting last 5 or 10 lines of `gui.log` into notification e-mail...! 

======

谢谢您的建议。）
事实上，我已经查看了 `debug/gui.log` ，但我认为要过滤掉这个日志文件以获取专门的信息非常困难。
例如，你可以看到日志数据并不是结构化的，如果我想提取 "哪个操作已完成 "或 "获取了多少 RMA-70-12"，就需要考虑多语言的问题。

此外，如果我运行多个 MAA 和应用程序播放器，要区分每个 MAA 实例将很困难。

但至少我可以尝试将最后 5 或 10 行 `gui.log` 插入通知电子邮件.....！
